Translate Tutorial (#83)
reactjs/es.reactjs.org: content/tutorial/tutorial.md

---
id: tutorial

> title: "Tutorial: Intro a React"

Reviewer: ```suggestion
title: "Tutorial: Introducción a React"
```

>Tip
>
>Este tutorial está diseñado para personas que prefieren **aprender haciendo**. Si tu prefieres aprender los conceptos desde el principio, revisa nuestra [guía paso a paso](/docs/hello-world.html). Podrías encontrar este tutorial y la guía, complementarias entre sí.

Este tutorial está dividido en varias secciones:

* [Configuración para el tutorial](#setup-for-the-tutorial) te dará un punto de partida para seguir el tutorial.
* [Visión general](#overview) te enseñará **los fundamentos** de React: componentes, props y estado.
* [Completando el juego](#completing-the-game) te enseñará **las técnicas más comunes** en desarrollo de React.
* [Agregando viaje en el tiempo](#adding-time-travel) te dará una **visión más profunda** de las fortalezas únicas de React.

No tienes que completar todas las secciones a la vez para obtener el valor de este tutorial. Prueba llegar tan lejos como puedas, incluso si es una o dos secciones.

> Está bien copiar y pegar el código mientras sigues el tutorial, pero te recomendamos que lo escribaas a mano. Esto te ayudará a desarrollar una memoria muscular y un entendimiento más sólido.

Reviewer: ```suggestion
Está bien copiar y pegar el código mientras sigues el tutorial, pero te recomendamos que lo escribas a mano. Esto te ayudará a desarrollar una memoria muscular y un entendimiento más sólido.
```

Puedes ver lo que construiremos aquí: **[Resultado Final](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**. Si el código no te hace sentido, o si no estás familiarizado con la sintaxis de código, ¡no te preocupes! El objetivo de este tutorial es ayudarte a entender React y su sintaxis.

Recomentamos que revises el juego de tic-tac-toe antes de continuar con el tutorial. Una de las características que notarás es que hay una lista enumerada a la derecha del tablero del jugador. Esta lista da una historia de todos los movimientos que han ocurrido en el juego, y se va actualizando conforme el juego progresa.

Puedes cerrar el juego de tic-tac-toe una vez que te familiarizaste con él. Empezaremos desde una plantilla más simple en este tutorial. Nuestro siguiente paso es configurarlo de tal forma que puedas empezar a construir el juego.

### Prerequisitos {#prerequisites}

Asumimos que tienes cierta familiaridad con HTML y JavaScript, pero deberías ser capaz de seguir adelante incluso si vienes de un lenguaje de programación diferente. También suponemos que estás familiarizado con conceptos de programación como funciones, objetos, arrays, y en menor medida, clases.

> Si necesitas revisar JavaScript, te recomendamos leer [esta guía](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript). Ten en cuenta que también usamos algunas características de ES6, una versión reciente de JavaScript. En este tutorial, estamos usando [funciones flecha](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [clases](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes), sentencias [`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) y [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const). Puedes usar el [Babel REPL](babel://es5-syntax-example) para revisar a qué código compila ES6.

Reviewer: ```suggestion
Si necesitas revisar JavaScript, te recomendamos leer [esta guía](https://developer.mozilla.org/es/docs/Web/JavaScript/A_re-introduction_to_JavaScript). Ten en cuenta que también usamos algunas características de ES6, una versión reciente de JavaScript. En este tutorial, estamos usando [funciones flecha](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [clases](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Classes), sentencias [`let`](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Statements/let) y [`const`](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Statements/const). Puedes usar el [Babel REPL](babel://es5-syntax-example) para revisar a qué código compila ES6.
```

¡Ésta es la forma más rápida de empezar!

Primero, abre este **[código inicial](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)** en una nueva pestaña. La nueva pestaña deberá mostrar un tablero vacío del juego de tic-tac-toe y código de React. Estaremos editando el código de React en este tutorial.

Ahora puedes saltarte a la segunda opción de configuración o ir a la sección de [visión general](#overview) para obtener una idea general de React.

### Opción de configuración 2: Entorno de desarrollo local {#setup-option-2-local-development-environment}

¡Ésta es completamente opcional y no es requeridad para éste tutorial!

<br>

<details>

<summary><b>Opcional: Instrucciones para seguir adelante localmente usando tu editor de texto preferido</b></summary>

Esta configuración requiere más trabajo pero te permite completar el tutorial usando un editor de tu elección. Aquí los pasos a seguir:

1. Asegúrate de tener una versión reciente de [Node.js](https://nodejs.org/en/) instalada.
2. Sigue las [instrucciones de instalación de Create React App](/docs/create-a-new-react-app.html#create-react-app) para hacer un nuevo proyecto.

```bash
npx create-react-app my-app
```

> 1. Elimina todos los archivos en la carpeta `src/` del nuevo proyecto.

Reviewer: ```suggestion
3. Elimina todos los archivos en la carpeta `src/` del nuevo proyecto.
```

> Nota:
>
>**No elimines la carpeta `src` por completo, solo los archivos de código fuente originales dentro de ella**. Reemplazaremos los archivos de código fuente por defecto con ejemplos para este proyecto en el siguiente paso.

```bash
cd my-app
cd src

# Si usas Mac ó Linux:
rm -f *

# Ó, si usas Windows:
del *

# Luego, regresa a la carpeta del proyecto
cd ..
```

> 1. Agrega un archivo llamado `index.css` en la carpeta `src/` con [este código CSS](https://codepen.io/gaearon/pen/oWWQNa?editors=0100).

Reviewer: ```suggestion
4. Agrega un archivo llamado `index.css` en la carpeta `src/` con [este código CSS](https://codepen.io/gaearon/pen/oWWQNa?editors=0100).
```

> 2. Agrega un archivo llamado `index.js` en la carpeta `src/` con [este código JS](https://codepen.io/gaearon/pen/oWWQNa?editors=0010).

Reviewer: ```suggestion
5. Agrega un archivo llamado `index.js` en la carpeta `src/` con [este código JS](https://codepen.io/gaearon/pen/oWWQNa?editors=0010).
```

> 3. Agrega estas 3 líneas en la parte superior del archivo `index.js` en la carpeta `src/`:

Reviewer: ```suggestion
6. Agrega estas 3 líneas en la parte superior del archivo `index.js` en la carpeta `src/`:
```

```js
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
```

Ahora, si tu ejecutas `npm start` en la carpeta del proyecto y abres `http://localhost:3000` en el navegador, deberías ver un campo de tic-tac-toe vacío.

Recomendamos seguir [estas instrucciones](https://babeljs.io/docs/editors/) para configurar el resaltado de sintaxis para tu editor.

</details>

### ¡Ayuda, estoy atorado! {#help-im-stuck}

Si te atoras, revisa los [recursos de soporte de la comunidad](/community/support.html). En particular, [el chat de Reactiflux](https://discord.gg/0ZcbPKXt5bZjGY5n) es una gran manera de obtener ayuda rápidamente. Si no recibes una respuesta, o sigues atorado, por favor crea un issue, y te ayudaremos.

## Visión General {#overview}

Ahora que está tu entorno configurado, ¡vamos a obtener una visión general de React!

### ¿Qué es React? {#what-is-react}

> React es una librería de JavaScript declarativa, eficiente y flexible para construir interfaces de usuario. Permite componer UIs complejas de pequeñas y aisladas piezas de código llamadas "componentes".

Reviewer: We're already using `IU` instead of `UI`, let's change it for now!

```suggestion
React es una librería de JavaScript declarativa, eficiente y flexible para construir interfaces de usuario. Permite componer IUs complejas de pequeñas y aisladas piezas de código llamadas "componentes".
```

);
```

[Ver la versión completa extendida.](babel://tutorial-expanded-version)

Si tienes curiosidad, `createElement()` está descrito en más detalle en la [referencia de la API](/docs/react-api.html#createelement), pero no lo usaremos en este tutorial. En cambio, seguiremos usando JSX.

JSX viene con todo el poder de JavaScript. Puedes poner *cualquier* expresión de JavaScript en el interior de las llaves dentro de JSX. Cada elemento de React es un objeto de JavaScript que puedes almacenar en una variable o pasar alrededor de tu programa.

> El componente anterior `ShoppingList` solo renderiza componentes pre-construidos del DOM como `<div />` y `<li />`. Pero, también puedes componener y renderizar componentes personalizados de React. Por ejemplo, ahora podemos referirmos al listado de compras completo escribiendo `<ShoppingList />`. Cada componente de React está encapsulado y puede operar independientemente; esto te permite construir UIs complejas desde componentes simples.

Reviewer: ```suggestion
El componente anterior `ShoppingList` solo renderiza componentes pre-construidos del DOM como `<div />` y `<li />`. Pero, también puedes componener y renderizar componentes personalizados de React. Por ejemplo, ahora podemos referirmos al listado de compras completo escribiendo `<ShoppingList />`. Cada componente de React está encapsulado y puede operar independientemente; esto te permite construir IUs complejas desde componentes simples.
```

}
```

Si hacemos click en un cuadrado ahora, deberíamos de obtener una alerta en nuestro navegador.

>Nota
>

> >Para continuar escribiendo código sin problemas y evitar el [confuso comportamiento de `this`](https://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/), vamos a usar la [sintaxis de funciones flecha](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) para manejar eventos aquí y más abajo:

Reviewer: ```suggestion
>Para continuar escribiendo código sin problemas y evitar el [confuso comportamiento de `this`](https://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/), vamos a usar la [sintaxis de funciones flecha](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Functions/Arrow_functions) para manejar eventos aquí y más abajo:
```

}
```

>Nota
>

> >En las [clases de JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes), necesitas siempre llamar `super` cuando defines el constructor de una subclase. Todas las clases de componentes de React que tienen un `constructor` deben empezar con una llamada a `super(props)`.

Reviewer: ```suggestion
>En las [clases de JavaScript](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Classes), necesitas siempre llamar `super` cuando defines el constructor de una subclase. Todas las clases de componentes de React que tienen un `constructor` deben empezar con una llamada a `super(props)`.
```

Ahora vamos a cambiar el método `render` de Square para mostrar el valor del estado actual cuando es clickeado:

* Reemplaza `this.props.value` por `this.state.value` dentro de la etiqueta `<button>`.
* Reemplaza el manejador de evento `() => alert()` por `() => this.setState({value: 'X'})`.
* Pon los props `className` y `onClick` en líneas separadas para mejor legibilidad.

> Luedo de estos cambios, la etiqueta `<button>` que es retornada del método `render` de Square se ve así:

Reviewer: ```suggestion
Luego de estos cambios, la etiqueta `<button>` que es retornada del método `render` de Square se ve así:
```

}
```

Llamando a `this.setState` desde el manejador `onClick` en el método `render` de Square, decimos a React que re-renderice el cuadrado siempre que su `<button>` es clickeado. Luego de la actualización, el `this.state.value` del cuadrado será `'X'`, así que veremos `X` en el tablero de juego. Si tu haces click en cualquier cuadrado, una `X` debería mostrarse en el mismo.

Cuando llamas `setState` en un componente, React actualiza automáticamente los componentes hijos dentro del mismo también.

**[Ver el código completo en este punto](https://codepen.io/gaearon/pen/VbbVLg?editors=0010)**

### Herramientas de desarrollo {#developer-tools}

La extensión de React Devtools para [Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) y [Firefox](https://addons.mozilla.org/en-US/firefox/addon/react-devtools/) te permite inspeccionar el árbol de componentes de React con tus herramientas de desarrollo del navegador.

<img src="../images/tutorial/devtools.png" alt="React Devtools" style="max-width: 100%">

> El React DevTools te permite revisar las props y el estado de tus componentes de React.

Reviewer: You're using `los props` across the whole document, let's keep it that way :)
```suggestion
React DevTools te permite revisar los props y el estado de tus componentes de React.
```

> Después de instalar React DevTools, puedes hacer click derecho en cualquier elemento de la página, click en "Inspect" para abrir las herramientas de desarrollo, y la pestaña de React aparecerá como la última pestaña a la derecha.

Reviewer: I think this is what Chrome says in Spanish
```suggestion
Después de instalar React DevTools, puedes hacer click derecho en cualquier elemento de la página, click en "Inspeccionar elemento" para abrir las herramientas de desarrollo, y la pestaña de React aparecerá como la última pestaña a la derecha.
```

3. Click en "Change View" y luego selecciona "Debug mode".
4. En la nueva pestaña que se abre, el devtools debería ahora tener una pestaña de React.

## Completando el juego {#completing-the-game}

Ahora tenemos los bloques de construcción básicos para nuestro juego tic-tac-toe. Para completar el juego, necesitamos alternar colocando "X" y "O" en el tablero, y necesitas una forma de determinar el ganador.

### Elevando el estado {#lifting-state-up}

Actualmente, cada componente Square mantiene el estado del juego. Para determinar un ganador, necesitamos mantener el valor de cada uno de los 9 cuadrador en un solo lugar.

Podemos pensar que el tablero debería solo preguntar a cada cuadrado por su estado. Aunque este enfoque es posible en React, te incentivamos a que no lo uses porque el código se vuelve difícil de ententer, susceptible a errores, y difícil de refactorizar. En su lugar, el mejor enfoque es almacenar el estado del juego en el componente padre Board en vez de cada componente Square. El componente Board puede decirle a cada cuadrado que mostrar pasándole un prop [tal cual hicimos cuando pasamos un número a cada cuadrado](#passing-data-through-props).

> **Para recopilar datos de múltiples hijos, o tener dos componentes hijos comunicados entre si, necesitas declarar el estado compartido en su componente padre. El componente padre puede pasar el estado hacia los hijos usando props; esto mantiene los componentes hijos sincronizados entre ellos y con su componente padre.**

Reviewer: ```suggestion
**Para recopilar datos de múltiples hijos, o tener dos componentes hijos comunicados entre sí, necesitas declarar el estado compartido en su componente padre. El componente padre puede pasar el estado hacia los hijos usando props; esto mantiene los componentes hijos sincronizados entre ellos y con su componente padre.**
```

  }

  render() {

>     const status = 'Siguiente jugador: X';

Reviewer: Please, roll back this change, we are not translating any code - we're just translating code comments.

]
```

El método `renderSquare` del componente Board actualmente se ve así:

```javascript
  renderSquare(i) {
    return <Square value={i} />;
  }
```

Al principio, [pasamos el prop `value`](#passing-data-through-props) desde el Board para mostrar los números de 0 a 8 en cada cuadrado. En un paso previo, reemplazamos los números con una marca "X" [determinado por el estado del propio Square](#making-an-interactive-component). Esto es porque el cuadrado actualmente ignora el prop `value` pasado por el Board.

Ahora usaremos el prop pasando el mecanismo otra vez. Modificaremos el Board para instruir cada Square acerca de su valor actual (`'X'`, `'O'`, ó `null`). Ya tenemos definido el arreglo `squares` en  el constructor del Board, y modificaremos el método `renderSquare` para que lo lea desde allí:

```javascript{2}
  renderSquare(i) {
    return <Square value={this.state.squares[i]} />;
  }
```

**[Ver el código completo en este punto](https://codepen.io/gaearon/pen/gWWQPY?editors=0010)**

Cada Square ahora recibirá un prop `value` que será `'X'`, `'O'`, ó `null` para cuadrados vacíos.

> Luego, necesitamos cambiar lo que sucede cuando un cuadrado es clickeado. El componente Board ahora mantiene qué cuadrados están rellenos. Necesitamos crear una forma para que el cuadrado actualiza el estado del componente Board. Debido a que el estado es considerado privado al componente que lo define, no podemos actualizar el stado de Board directamente desde Square.

Reviewer: ```suggestion
Luego, necesitamos cambiar lo que sucede cuando un cuadrado es clickeado. El componente Board ahora mantiene qué cuadrados están rellenos. Necesitamos crear una forma para que el cuadrado actualiza el estado del componente Board. Debido a que el estado es considerado privado al componente que lo define, no podemos actualizar el estado de Board directamente desde Square.
```

}
```

Cuando un cuadrado es clickeado, la función `onClick` provista por el componente Board es llamada. Aquí un repaso de cómo esto fue logrado:

1. El prop `onClick` en el componente pre-construido del DOM `<button>` le dice a React para establecer un escuchador del evento click.
2. Cuando el botón es clickeado, React llamará al manejador de evento `onClick` que está definido en el método `render()` de Square.
3. Este manejador de evento llama a `this.props.onClick()`. El prop `onClick` del componente Square fue especificado por el componente Board.
4. Debido a que el Board pasó `onClick={() => this.handleClick(i)}` a Square, el componente Square llama a `this.handleClick(i)` cuando es clickeado.
5. No tenemos definido el método `handleClick()` aun, así que nuestro código falla.

>Nota
>

> >El atributo `onClick` del elemento `<button>` del DOM tiene un significado especial para React porque es un componente pre-construido. PAra componentes personalizados como Square, la nomenclatura la decides tú. Podemos nombrar el prop `onClick` de Square o el método `handleClick` de Board diferente. En React, sin embargo, es una convención usar los nombres `on[Evento]` para props que representan eventos y `handle[Event]` para los métodos que manejan los eventos.

Reviewer: ```suggestion
>El atributo `onClick` del elemento `<button>` del DOM tiene un significado especial para React porque es un componente pre-construido. Para componentes personalizados como Square, la nomenclatura la decides tú. Podemos nombrar el prop `onClick` de Square o el método `handleClick` de Board diferente. En React, sin embargo, es una convención usar los nombres `on[Evento]` para props que representan eventos y `handle[Event]` para los métodos que manejan los eventos.
```

> Cuando intentamos clickear un cuadrado, deberíamos obtener un eerror porque no hemos definido `handleClick` aun. Vamos ahora a agregar `handleClick` a la clase Board:

Reviewer: ```suggestion
Cuando intentamos clickear un cuadrado, deberíamos obtener un error porque no hemos definido `handleClick` aun. Vamos ahora a agregar `handleClick` a la clase Board:
```

Debido a que el componente Square ahora no mantiene estado, los componentes Square reciben valores del  componente Board e informan al mismo cuando son clickeados. En términos de React, los componentes Square ahora son **componentes controlados**. El componente Board tiene control completo sobre ellos.

Notar cómo en `handleClick`, llamamos `.slice()` para crear una copia del array de `squares` para modificarlo en vez de modificar el array existente. Ahora explicareomos porqué crear una copia del array `squares` en la siguiente sección.

### ¿Por qué es importante la Inmutabilidad? {#why-immutability-is-important}

En el ejemplo de código anterior, sugerimos que uses el operador `.slice()` para crear una copia del array de `squares` para modificar en vez de modificar el array existente. Ahora discutiremos inmutabilidad y porqué es importante aprenderlo.

> Hay generalmente dos enfoques para cambiar datos. El primer enfoque es *mutar* los datos directamente cambienado sus valores. El segundo enfoque es reemplazar los datos con una nueva copia que tiene los cambios deseados.

Reviewer: ```suggestion
Hay generalmente dos enfoques para cambiar datos. El primer enfoque es *mutar* los datos directamente cambiando sus valores. El segundo enfoque es reemplazar los datos con una nueva copia que tiene los cambios deseados.
```

#### Cambio de datos con mutación {#data-change-with-mutation}
```javascript
var player = {score: 1, name: 'Jeff'};
player.score = 2;
// Ahora `player` es {score: 2, name: 'Jeff'}
```

#### Cambio de datos sin mutación {#data-change-without-mutation}
```javascript
var player = {score: 1, name: 'Jeff'};

var newPlayer = Object.assign({}, player, {score: 2});
// Ahora `player` no ha cambiado, pero `newPlayer` es {score: 2, name: 'Jeff'}

// O si usas la sintaxis propuesta de propagación de objeto, puedes escribir:
// var newPlayer = {...player, score: 2};
```

El resultado final es el mismo, pero al no mutar (o cambiar los datos subyacentes) directamente, obtenemos muchos beneficios descritos a continuación

#### Funcionalidades complejas se vuelven simples {#complex-features-become-simple}

> Inmutabilidad hace que funcionalidades complejas sean mucho más fácil de implementar. Luego en este tutorial, implementaremos una funcionalidad de "viaje en el tiempo" que nos permite repasar el historial del juego tic-tac-toe y "volver" a movimientos previos. Esta funcionalidad no es específica de juegos, una habilidad de deshacer y rehacer ciertas acciones es un requerimiento común en aplicaciones. Evitar la mutación de datos directamente nos permite mantener intacto versiones previas del historial del juego, y reusarlas luego.

Reviewer: ```suggestion
La Inmutabilidad hace que funcionalidades complejas sean mucho más fácil de implementar. Luego en este tutorial, implementaremos una funcionalidad de "viaje en el tiempo" que nos permite repasar el historial del juego tic-tac-toe y "volver" a movimientos previos. Esta funcionalidad no es específica de juegos, una habilidad de deshacer y rehacer ciertas acciones es un requerimiento común en aplicaciones. Evitar la mutación de datos directamente nos permite mantener intacto versiones previas del historial del juego, y reusarlas luego.
```

  }
```

Con este cambio, "X"s y "O"s pueden tomar turnos. También vamos a cambiar el texto de "status" en el `render` del Board para que muestre qué jugador tiene el siguiente turno:

```javascript{2}
  render() {

>     const status = 'Siguiente jugador: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
    const status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

  }

  render() {

>     const status = 'Siguiente jugador: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
    const status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

}
```

Llamaremos a `calculateWinner(squares)` en el método `render` del componente Board para revisar si un jugador a ganado. Si un jugador ha ganado, podemos mostrar un texto como: "Ganador: X" o "Ganador: O". Reemplazaremos la declaración del `status` en el método `render` de Board con este código:

```javascript{2-8}
  render() {
    const winner = calculateWinner(this.state.squares);
    let status;
    if (winner) {

>       status = 'Ganador: ' + winner;

Reviewer: ```suggestion
      status = 'Winner: ' + winner;
```

    } else {

>       status = 'Siguiente jugador: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

}
```

> Llamaremos a `calculateWinner(squares)` en el método `render` del componente Board para revisar si un jugador a ganado. Si un jugador ha ganado, podemos mostrar un texto como: "Winner: X" o "Winner: O". Reemplazaremos la declaración del `status` en el método `render` de Board con este código:

Reviewer 2: **ha** en _a ganado_, lleva hache (porque es de haber)

Author: Good catch!

Author: Done!

    const winner = calculateWinner(this.state.squares);
    let status;
    if (winner) {

>       status = 'Ganador: ' + winner;

Reviewer: ```suggestion
      status = 'Winner: ' + winner;
```

    } else {

>       status = 'Siguiente jugador: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

    let status;
    if (winner) {

>       status = 'Ganador: ' + winner;

Reviewer: ```suggestion
      status = 'Winner: ' + winner;
```

    } else {

>       status = 'Siguiente jugador: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

  }
```

>Nota
>
>A diferencia del método `push()` de los arrays que debes estar más familiarizado, el método `concat()` no mutal el array original, por eso lo preferimos.

En este punto, el componente Board solo necesita los métodos `renderSquare` y `render`. El estado del juego y el método `handleClick` deberían estar en el componente Game.

**[Ver el código completo en este punto](https://codepen.io/gaearon/pen/EmmOqJ?editors=0010)**

### Mostrando los movimientos anteriores {#showing-the-past-moves}

Desde que grabamos el historial del juego tic-tac-toe, ahora podemos mostrarlo al jugador como una lista de movimientos anteriores.

Aprendimos antes que los elementos de React son objetos de primera clase en JavaScript; así que podemos pasarlo alrededor de nuestras aplicaciones. Para renderizar múltiples elementos en React, podemos usar un array de elementos de React.

> En JavaScript, los arrays tienen un [método `map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) que es comúnmente usado para mapear datos a otros datos, por ejemplo:

Reviewer: ```suggestion
En JavaScript, los arrays tienen un [método `map()`](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Array/map) que es comúnmente usado para mapear datos a otros datos, por ejemplo:
```

    const moves = history.map((step, move) => {
      const desc = move ?

>         'Ir al movimiento #' + move :

Reviewer: Please, roll this back, we're not translating code fragments.

Thanks!

Si el key no está especificado, React presentará una advertencia y usará el índice del array como índice por defecto. Usando el índice del array como un key es problemático cuando intentas reordenar los elementos de una lista ó insertar/eliminar elementos de la lista. Pasar explícitamente `key={i}` silencia la advertencia pero tiene los mismos problemas que los índices del array y no es recomendado en la mayoría de los casos.

Los keys no necesitan ser globalmente únicos; solo necesitan ser únicos entre componentes y sus hermanos..

### Implementando viaje en el tiempo {#implementing-time-travel}

En el historial del juego de tic-tac-toe, cada movimiento anterior tiene un ID único asociado; es el número secuencial del movimiento. Los movimientos nunca son reordenados, eliminados, ó insertados en el medio, así que es seguro usar los índices del movimiento como un key.

En el método `render` del componente Game, podemos agregar el key como `<li key={move}` la advertencia de React debería desaparecer:

```js{6}
    const moves = history.map((step, move) => {
      const desc = move ?

>         'Ir al movimiento #' + move :

Reviewer: Roll this code back also, please.

* Te permite jugar tic-tac-toe,
* Indica cuando un jugador ha ganado el juego,
* Almacena el historial del juego como va progresando,
* Permite a los jugadores revisar el historial del juego y ver versiones anteriores del tablero de juego.

¡Buen trabajo! Esperamos que ahora te sientas que tienes un entendimiento descente sobre cómo funciona React.

Revisa el resultado final aquí: **[Resultado final](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**.

Si tienes un tiempo extra o quieres practicar tus nuevas habilidades de React, aquí algunas ideas de mejoras que puedes hacer al juego de tic-tac-toe, las cuales están listadas en orden de dificultad creciente:

1. Muestra la ubicación para cada movimiento en el formato (columna, fila) en la lista del historial de movimientos.
2. Convierte en negrita el elemento actualmente seleccionado en la lista de movimientos.
3. Reescribe el Board para usar 2 ciclos para hacer los cuadrados en vez de escribirlos a mano.
4. Agrega un botón de switch que te permita ordenar los movimientos en orden ascendente o descendente.
5. Cuando alguien gana, resalta los 3 cuadrados que hicieron que gane.

> 6. Cuando nadie gane, muestra un mensaje acerca de que el resultado es un empate.

Reviewer: ```suggestion
6. Cuando nadie gana, muestra un mensaje acerca de que el resultado es un empate.
```
